Doberman/AlarmNode.py

```diff
@@ -0,0 +1,124 @@
+        self.escalation_config = kwargs['escalation_config']
+        self.escalation_level = 0
+        self.base_level = kwargs['base_level']
+        self.auto_silence_duration = kwargs.get('silence_duration', 5*60)
+        self.alarm_start = None
+        self.hash = None
+
+    def escalate(self):
+        """
+        Do we escalate? This function decides this
+        """
+        if self.alarm_start is None:
+            # not sure how we got here, but there's nothing to escalate
+            return
+        time_since_start = time.time() - self.alarm_start
+        if time_since_start > self.escalation_config[max(self.base_level + self.escalation_level, len(self.escalation_config)-1)]:
+            self.logger.warning((f'{self.name} at level {self.base_level}/{self.escalation_level} '
+                f'since {time_since_start//60} minutes, time to escalate (hash {self.hash})'))
+            self.escalation_level += 1
+            self.escalation_level = min(len(self.escalation_config) - self.base_level - 1, self.escalation_level)
+            self.alarm_start = time.time()  # reset start time so we don't escalate again immediately
+
+    def reset_alarm(self):
+        """
+        Resets the cached alarm state
+        """
+        if self.hash is not None:
+            self.logger.info(f'{self.name} resetting alarm {self.hash}')
+            self.hash = None
+            # TODO should we also reset self.alarm_start?
+        self.escalation_level = 0
+
+    def log_alarm(self, msg, ts=None):
+        """
+        Let the outside world know that something is going on
+        """
+        if not self.is_silent:
+            self.logger.debug(msg)
+            if self.hash is None:
+                self.hash = Doberman.utils.make_hash(ts or time.time(), self.pipeline.name)
+                self.alarm_start = ts or time.time()
+                self.logger.warning(f'{self.name} beginning alarm with hash {self.hash}')
+            self.escalate()
+            self._log_alarm(msg, self.pipeline.name, self.hash, self.base_level, self.escalation_level)
+            self.pipeline.silence_for(self.auto_silence_duration)
+        else:
+            self.logger.error(msg)
+
+class SensorRespondingAlarm(Doberman.InfluxSourceNode, AlarmNode):
+    """
+    A simple alarm that makes sure the spice is flowing
+    """
+    def setup(self, **kwargs):
+        super().setup(**kwargs)
+        self.late_counter = 0
+        self.late_threshold = 3 # TODO config-ize
+
+    def reset_alarm(self):
+        super().reset_alarm()
+        self.late_counter = 0
+
+    def get_package(self):
+        try:
+            ret = super().get_package()
+            self.reset_alarm()
+            return ret
+        except ValueError as e:
+            self.late_counter += 1
+            if self.late_counter >= self.late_threshold:
+                self.log_alarm(f"Is {self.sensor} responding correctly? We haven't gotten a new value recently")
```

Comment:
Add `self.late_threshold` to the message

Reply:
Actually, is this alarm sent only if the problem is with database access? In which case maybe adjust the message to make that clearer

Reply:
The base class (InfluxSourceNode, in this case) raises a `ValueError` if it gets the same value twice from the database, because this corrupts the buffer further down the pipeline and leads to a multitude of errors. Here we just catch this error, increment up the counter because we missed one value, and re-raise. If the problem is with the database we'd be getting a variety of other messages as well.